### PR TITLE
[nereids](mtmv)Support rewrite by mv nested materialized view

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Support query rewritting by nested materialized view.
Such as `inner_mv` def is as following

>             select
>             l_linenumber,
>             o_custkey,
>             o_orderkey,
>             o_orderstatus,
>             l_partkey,
>             l_suppkey,
>             l_orderkey
>             from lineitem
>             inner join orders on lineitem.l_orderkey = orders.o_orderkey;

the mv1_0 def is as following:

>             select
>             l_linenumber,
>             o_custkey,
>             o_orderkey,
>             o_orderstatus,
>             l_partkey,
>             l_suppkey,
>             l_orderkey,
>             ps_availqty
>             from inner_mv
>             inner join partsupp on l_partkey = ps_partkey AND l_suppkey = ps_suppkey;

for the following query, both inner_mv and mv1_0 can be successful when query rewritting by materialized view，and cbo will chose `mv1_0` finally.

>            select lineitem.l_linenumber
>             from lineitem
>             inner join orders on l_orderkey = o_orderkey
>             inner join partsupp on  l_partkey = ps_partkey AND l_suppkey = ps_suppkey
>             where o_orderstatus = 'o' AND l_linenumber in (1, 2, 3, 4, 5)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Review by keanji-x — APPROVED** [review]

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewRule.java [hunk 1]
```diff
@@ -85,52 +87,62 @@ public abstract class AbstractMaterializedViewRule implements ExplorationRuleFac
...
-                );
-            }
-            return validQueryStructInfos;
-        }
-        validQueryStructInfos.add(queryStructInfo);
-        return validQueryStructInfos;
+    protected List<StructInfo> getValidQueryStructInfos(Plan queryPlan, CascadesContext cascadesContext,
+            BitSet materializedViewTableSet) {
+        return MaterializedViewUtils.extractStructInfo(queryPlan, cascadesContext, materializedViewTableSet)
+                .stream()
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewRule.java:133]: use for loop

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializedViewUtils.java [hunk 2]
```diff
@@ -132,18 +139,46 @@ public static boolean containTableQueryOperator(Plan analyzedPlan) {
...
+            StructInfoMap structInfoMap = ownerGroup.getstructInfoMap();
+            structInfoMap.refresh(ownerGroup);
+            Set<BitSet> queryTableSets = structInfoMap.getTableMaps();
+            if (!queryTableSets.isEmpty()) {
+                return queryTableSets.stream()
+                        // Just construct the struct info which mv table set contains all the query table set
+                        .filter(queryTableSet -> materializedViewTableSet.isEmpty()
+                                || StructInfo.containsAll(materializedViewTableSet, queryTableSet))
+                        .map(tableMap -> structInfoMap.getStructInfo(tableMap, tableMap, ownerGroup, plan))
+                        .collect(Collectors.toList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializedViewUtils.java:156]: use for loop

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializedViewUtils.java [hunk 3]
```diff
@@ -132,18 +139,46 @@ public static boolean containTableQueryOperator(Plan analyzedPlan) {
...
+     * should be different
+     */
+    public static Plan generateMvScanPlan(MTMV materializedView, CascadesContext cascadesContext) {
+        LogicalOlapScan mvScan = new LogicalOlapScan(
+                cascadesContext.getStatementContext().getNextRelationId(),
+                materializedView,
+                ImmutableList.of(materializedView.getQualifiedDbName()),
+                // this must be empty, or it will be used to sample
+                Lists.newArrayList(),
+                Lists.newArrayList(),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializedViewUtils.java:176]: use ImmutalbeList.of()

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/ExpressionLineageReplacer.java [hunk 4]
```diff
@@ -43,12 +45,12 @@
  * Get from rewrite plan and can also get from plan struct info, if from plan struct info it depends on
  * the nodes from graph.
  */
-public class ExpressionLineageReplacer extends DefaultPlanVisitor<Expression, ExpressionReplaceContext> {
+public class ExpressionLineageReplacer extends DefaultPlanVisitor<Void, ExpressionReplaceContext> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/ExpressionLineageReplacer.java:48]: replacer return Void is weird

**Review by morrySnow — COMMENTED** [review]
update all stream api to for loop

**Review by morrySnow — APPROVED** [review]

**Pull request merged**